### ISSUE TITLE
[WIDGET] taskbar

### ISSUE THREAD

**Issue by godalming123** [elkowar/eww#414]
### Description of the widget

A basic taskbar widget like that in windows or waybar.

### Implementation proposal

This would use the box widget as well as smaller buttons with an image and label.

### Example usage

```
(taskbar :format "icon name"
              :icon-size 15
              :tooltip-format ""
              :on-click "activate"
              :on-middle-click "close"
              :on-right-click "new-instance"
              :ignore "alacritty"
              :direction "h"
```

**Comment by elkowar** [comment]
Given that that _heavily_ depends on your display server, and is _really_ hard to do in a way that is flexible enough to fit with the spirit of eww, I'll probably say a builtin Taskbar widget is largely out of scope for now. (not 100% sure on that, of course) 

In the future there might be some sort of plugin system, or some other cleaner way to have external extensions to eww that can directly interact with gtk -- with those, such a widget might be a thing again.

Until then, do know that a Taskbar widget should already generally be achievable via some scripts, literal, etc - not super pretty, but possible

**Comment by godalming123** [comment]
Do you have any help in finding those scripts I had a polybar windows script on xorg but that doesnt work on wayland is there any tool/script that works for wayland?

**Comment by VuiMuich** [comment]
You might want to search the issues, there has been some discussion about using `stalonetray` or `trayer` iirc.

**Comment by elkowar** [comment]
> Do you have any help in finding those scripts I had a polybar windows script on xorg but that doesnt work on wayland is there any tool/script that works for wayland?

I don't know of any specific tools, no. On Wayland these things tend to be a lot more complex, as I understand it, as Wayland gives applications far less access to information about, well, anything (for security reasons). 
My assumption is that such scripts would have to be compositor dependant, unless there is some well supported extension protocol for that info - idk about that.

Should most definitely not be _tooooo_ hard to do yourself, depending on your compositor, ofc

**Comment by Electrostasy** [comment]
To add to this from the Wayland side, something like [what is proposed in this wlrctl issue](https://todo.sr.ht/~brocellous/wlrctl/11) would definitely help script a taskbar, but it doesn't seem to be a priority for now. I haven't been able to find a tool that can be molded into a functional taskbar yet besides wlrctl (if the feature proposed in the issue could be implemented), unfortunately.

Waybar and tools similar to wlrctl use the [wlr-foreign-toplevel-management-unstable](https://gitlab.freedesktop.org/wlroots/wlr-protocols/-/blob/master/unstable/wlr-foreign-toplevel-management-unstable-v1.xml) protocol extension to receive and communicate toplevel (window) events with the compositor. It seems most compositors currently actively developed support it (not sure about mutter and kwin, though). Waybar's "wlr/taskbar" module seems to maintain its own list of opened toplevels, handles compositor toplevel events, asks the compositor to minimize/maximize/close toplevels etc.

There are more tools that can list opened Wayland toplevels, such as [Waylevel](https://git.sr.ht/~shinyzenith/waylevel) and [lswt](https://git.sr.ht/~leon_plickat/lswt), though monitoring toplevel events seems out of scope for them.

**Comment by Bavalionprolate** [comment]
> ### Описание виджета
> 
> Базовый виджет панели задач, такой как в окнах или путевой панели.
> ### Предложение по реализации
> 
> Это будет использовать виджет коробки, а также меньшие кнопки с изображением и меткой.
> ### Пример использования
> 
> ```
> (taskbar :format "icon name"
>               :icon-size 15
>               :tooltip-format ""
>               :on-click "activate"
>               :on-middle-click "close"
>               :on-right-click "new-instance"
>               :ignore "alacritty"
>               :direction "h"
> ```
> 
> ### Дополнительный контекст
Good idea to have a built-in taskbar widget, I'm literally looking for ways to write myself a taskbar widget for my rice myself. I saw a rice on reddit not too long ago where a person replicated the win11 design and recreated literally everything with eww (and taskbar, start menu, gamebar and more), I can attach the link below for you. Hope this helps you in some way, as I can't figure out the colossal amount of scripts and how they work.

link - https://www.reddit.com/r/unixporn/comments/13zdhqd/hyprland_windows_rice_with_too_much_eww_with_blur/

**Comment by siduck** [comment]
it'd be great if there were some initially supported compositors ( xorg's pretty easy ) and provide a magic variable ( which is a json arr of objects ) and each object has useful info about each window

example

![image](https://github.com/elkowar/eww/assets/59060246/546ba924-28a3-4817-9a16-e465d4437d1c)

**Comment by siduck** [comment]
I managed to make a taskbar widget in my bar, i needed a lot of scripting & have to use wmctrl for click actions and using xprop to for capturing xorg events

Rofi to show menu of multiple window titles of the same app just above the app's icon


https://github.com/elkowar/eww/assets/59060246/94d8d755-b860-48b6-b77f-a37dad601623

**Comment by RpxdYTX** [comment]
@siduck is this for x11 or wayland? Nonetheless, could you share the code?

**Comment by siduck** [comment]
> @siduck is this for x11 or wayland? Nonetheless, could you share the code?

the code is private and its for x11. I think its better for eww to implement it i.e for ewmh | wlroots

**Comment by Light034** [comment]
Hello, I made a standalone program for wayland compositors that uses [wlr-foreign-toplevel-management-unstable](https://gitlab.freedesktop.org/wlroots/wlr-protocols/-/blob/master/unstable/wlr-foreign-toplevel-management-unstable-v1.xml) and listens for events and updates the output in real time in json format. Inside my repo I also included an example eww configuration showing how this can be implemented in eww to display a very basic taskbar with limited functionality. 

I tested it with Wayfire but I think other compositors should work aswell. 

Here is the link for my program:
[wlr-apps](https://github.com/Light034/wlr-apps) 

And a screenshot of how it looks with the provided eww example:
![Image](https://github.com/user-attachments/assets/883c890d-4b32-4ea0-92e9-a7fd329013f2)
